### PR TITLE
Add EntityEvent.Constructing and EntityEvent.JoinWorld

### PR DESCRIPTION
**Pull request by gegy1000**

### CODE REVIEW

File: patches/net/minecraft/entity/Entity.java.patch [hunk 1]
```diff
@@ -1,6 +1,23 @@
 --- /net/minecraft/entity
 +++ /net/minecraft/entity
-@@ -308,6 +308,7 @@
+@@ -28,8 +28,6 @@
+ import net.minecraft.crash.ICrashReportDetail;
+ import net.minecraft.enchantment.EnchantmentHelper;
+ import net.minecraft.enchantment.EnchantmentProtection;
+-import net.minecraft.entity.EntityList;
+-import net.minecraft.entity.EntityLivingBase;
```
Comment (iLexiconn) [inline patches/net/minecraft/entity/Entity.java.patch]: no

File: patches/net/minecraft/world/World.java.patch [hunk 2]
```diff
@@ -1,6 +1,15 @@
 --- /net/minecraft/world
 +++ /net/minecraft/world
-@@ -1673,7 +1673,8 @@
+@@ -832,6 +832,8 @@
+                 this.updateAllPlayersSleepingFlag();
+             }
+ 
++            if (!xyz.openmodloader.OpenModLoader.INSTANCE.getEventBus().post(new xyz.openmodloader.event.impl.EntityEvent.JoinWorld(var1)) && !var4) return false;
++
```
Comment (iLexiconn) [inline patches/net/minecraft/world/World.java.patch]: no

File: patches/net/minecraft/world/World.java.patch [hunk 3]
```diff
@@ -10,11 +19,34 @@
...
+@@ -2273,10 +2277,11 @@
+     }
+ 
+     public void loadEntities(Collection<Entity> var1) {
+-        this.loadedEntityList.addAll(var1);
+-
+         for(Entity var3 : var1) {
+-            this.onEntityAdded(var3);
++            if (xyz.openmodloader.OpenModLoader.INSTANCE.getEventBus().post(new xyz.openmodloader.event.impl.EntityEvent.JoinWorld(var3))) {
++                this.loadedEntityList.add(var3);
```
Comment (iLexiconn) [inline patches/net/minecraft/world/World.java.patch]: make a bit smaller?

File: src/main/java/xyz/openmodloader/event/impl/EntityEvent.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+         */
+        public Constructing(Entity entity) {
+            super(entity);
+        }
+    }
+
+    /**
+     * This event is fired when an entity joins the world.
+     */
+    public static class JoinWorld extends EntityEvent {
```
Comment (iLexiconn) [inline src/main/java/xyz/openmodloader/event/impl/EntityEvent.java]: rename to Join

File: patches/net/minecraft/entity/Entity.java.patch (hunk 1 above)
Comment (shadowfacts) [inline patches/net/minecraft/entity/Entity.java.patch]: Lexi means no import changes

File: patches/net/minecraft/world/World.java.patch (hunk 2 above)
Comment (shadowfacts) [inline patches/net/minecraft/world/World.java.patch]: Here she means no adding empty lines

File: patches/net/minecraft/world/World.java.patch [hunk 5]
```diff
@@ -10,11 +19,34 @@
...
 +        xyz.openmodloader.OpenModLoader.INSTANCE.getEventBus().post(new xyz.openmodloader.event.impl.UpdateEvent.WorldUpdate(this));
          this.updateWeather();
      }
  
+@@ -2273,10 +2277,11 @@
+     }
+ 
+     public void loadEntities(Collection<Entity> var1) {
+-        this.loadedEntityList.addAll(var1);
+-
```
Comment (shadowfacts) [inline patches/net/minecraft/world/World.java.patch]: You're removing an empty line

**Comment by gegy1000** [comment]
Woah, that was fast. I'll fix the problems.

File: patches/net/minecraft/world/World.java.patch [hunk 6]
```diff
@@ -10,11 +19,34 @@
...
+         for(Entity var3 : var1) {
+-            this.onEntityAdded(var3);
++            if (xyz.openmodloader.OpenModLoader.INSTANCE.getEventBus().post(new xyz.openmodloader.event.impl.EntityEvent.JoinWorld(var3))) {
```
Comment (shadowfacts) [inline patches/net/minecraft/world/World.java.patch]: You can make the patch smaller by doing something like this:

```
if (!post(new Event()) continue;
this.loadedEntityList.add(var3);
```

Decreasing the patch size by 2 lines

File: patches/net/minecraft/entity/Entity.java.patch [hunk 7]
```diff
@@ -1,23 +1,14 @@
 --- /net/minecraft/entity
 +++ /net/minecraft/entity
-@@ -28,8 +28,6 @@
- import net.minecraft.crash.ICrashReportDetail;
- import net.minecraft.enchantment.EnchantmentHelper;
- import net.minecraft.enchantment.EnchantmentProtection;
--import net.minecraft.entity.EntityList;
--import net.minecraft.entity.EntityLivingBase;
```
Comment (iLexiconn) [inline patches/net/minecraft/entity/Entity.java.patch]: Still imports :P

Comment (gegy1000) [inline patches/net/minecraft/entity/Entity.java.patch]: Those were the imports you sent me. :P

Comment (shadowfacts) [inline patches/net/minecraft/entity/Entity.java.patch]: @iLexiconn No he removed the imports, that's just the git diff.

File: patches/net/minecraft/world/World.java.patch [hunk 8]
```diff
@@ -1,6 +1,43 @@
...
+-import net.minecraft.world.Explosion;
+-import net.minecraft.world.GameRules;
+-import net.minecraft.world.IBlockAccess;
+-import net.minecraft.world.IWorldEventListener;
+-import net.minecraft.world.MinecraftException;
+-import net.minecraft.world.NextTickListEntry;
+-import net.minecraft.world.WorldProvider;
+-import net.minecraft.world.WorldSavedData;
+-import net.minecraft.world.WorldSettings;
+-import net.minecraft.world.WorldType;
```
Comment (shadowfacts) [inline patches/net/minecraft/world/World.java.patch]: These, however, are still imports being removed.

File: patches/net/minecraft/world/World.java.patch [hunk 9]
```diff
@@ -1,6 +1,43 @@
...
+ import net.minecraft.world.biome.Biome;
+ import net.minecraft.world.biome.BiomeProvider;
+ import net.minecraft.world.border.WorldBorder;
+@@ -71,6 +58,7 @@
+ import net.minecraft.world.storage.MapStorage;
+ import net.minecraft.world.storage.WorldInfo;
+ import net.minecraft.world.storage.loot.LootTableManager;
++import xyz.openmodloader.event.impl.EntityEvent;
```
Comment (shadowfacts) [inline patches/net/minecraft/world/World.java.patch]: And added, in this case.

File: patches/net/minecraft/world/World.java.patch [hunk 10]
```diff
@@ -10,11 +47,34 @@
...
      }
  
+@@ -2273,12 +2265,12 @@
+     }
+ 
+     public void loadEntities(Collection<Entity> var1) {
+-        this.loadedEntityList.addAll(var1);
+ 
+         for(Entity var3 : var1) {
++            if (xyz.openmodloader.OpenModLoader.INSTANCE.getEventBus().post(new EntityEvent.Join(var3))) continue;
```
Comment (shadowfacts) [inline patches/net/minecraft/world/World.java.patch]: There should be a `!` at the beginning of this expression. The event bus returns `true` if all the handlers ran successfully and `false` if the event was canceled.

File: patches/net/minecraft/world/World.java.patch [hunk 11]
```diff
@@ -10,11 +47,34 @@
...
++            this.loadedEntityList.add(var3);
+             this.onEntityAdded(var3);
+         }
+-
```
Comment (shadowfacts) [inline patches/net/minecraft/world/World.java.patch]: You're removing an empty line.

File: patches/net/minecraft/world/WorldServer.java.patch [hunk 12]
```diff
@@ -1,6 +1,43 @@
...
+-import net.minecraft.world.WorldSettings;
+-import net.minecraft.world.WorldType;
+ import net.minecraft.world.biome.Biome;
+ import net.minecraft.world.biome.BiomeProvider;
+ import net.minecraft.world.chunk.Chunk;
+@@ -90,6 +76,7 @@
+ import net.minecraft.world.storage.loot.LootTableManager;
+ import org.apache.logging.log4j.LogManager;
+ import org.apache.logging.log4j.Logger;
++import xyz.openmodloader.event.impl.EntityEvent;
```
Comment (shadowfacts) [inline patches/net/minecraft/world/WorldServer.java.patch]: More import changes

File: patches/net/minecraft/world/World.java.patch [hunk 13]
```diff
@@ -10,11 +18,34 @@
...
+ 
+     public void loadEntities(Collection<Entity> var1) {
+-        this.loadedEntityList.addAll(var1);
+ 
+         for(Entity var3 : var1) {
++            if (xyz.openmodloader.OpenModLoader.INSTANCE.getEventBus().post(new xyz.openmodloader.event.impl.EntityEvent.Join(var3))) continue;
++            this.loadedEntityList.add(var3);
+             this.onEntityAdded(var3);
+         }
+-
```
Comment (shadowfacts) [inline patches/net/minecraft/world/World.java.patch]: You're still removing an empty line here

**Comment by gegy1000** [comment]
It's all fixed. _I'm very good at this._ Sorry.

**Comment by gegy1000** [comment]
Whoops, I see I still forgot to add the ! to loadEntities. Sorry for being such an inconvenience.

**Comment by iLexiconn** [comment]
You still broke less than I did the last few days, so that's a good thing.

**Pull request merged**